atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/monitoring/TimestampTrackerTest.java

```diff
@@ -0,0 +1,161 @@
+import static org.mockito.Mockito.when;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.junit.Test;
+
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.Gauge;
+import com.palantir.atlasdb.cleaner.Cleaner;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
+import com.palantir.lock.v2.TimelockService;
+
+public class TimestampTrackerTest {
+    private static final long ONE = 1L;
+    private static final long TEN = 10L;
+    private static final long FORTY_TWO = 42L;
+
+    private static final String IMMUTABLE_TIMESTAMP_NAME = "timestamp.immutable";
+    private static final String FRESH_TIMESTAMP_NAME = "timestamp.fresh";
+    private static final String UNREADABLE_TIMESTAMP_NAME = "timestamp.unreadable";
+    private static final String FAKE_METRIC = "metric.fake";
+
+    private static final long CACHE_RETRIGGER_NANOS = TimestampTracker.CACHE_INTERVAL.toNanos() + 1L;
+
+    private final TimelockService timelockService = mock(TimelockService.class);
+    private final Cleaner cleaner = mock(Cleaner.class);
+
+    private final Clock mockClock = mock(Clock.class);
+
+    @Test
+    public void throwsIfAddingMetricsMultipleTimes() {
+        try (TimestampTracker tracker = new TimestampTracker(Clock.defaultClock())) {
+            tracker.registerTimestampForTracking(FAKE_METRIC, () -> 1L);
+            assertThatThrownBy(() -> tracker.registerTimestampForTracking(FAKE_METRIC, () -> 2L))
+                    .isInstanceOf(IllegalStateException.class);
+        }
+    }
+
+    @Test
+    public void defaultTrackerGeneratesTimestampMetrics() {
+        try (TimestampTracker ignored = TimestampTracker.createWithDefaultTrackers(timelockService, cleaner)) {
+            assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                    .contains(buildFullyQualifiedMetricName(IMMUTABLE_TIMESTAMP_NAME))
+                    .contains(buildFullyQualifiedMetricName(FRESH_TIMESTAMP_NAME))
+                    .contains(buildFullyQualifiedMetricName(UNREADABLE_TIMESTAMP_NAME));
+        }
+    }
+
+    @Test
+    public void defaultTrackersDelegateToRelevantComponents() {
+        try (TimestampTracker ignored = TimestampTracker.createWithDefaultTrackers(timelockService, cleaner)) {
+            when(timelockService.getImmutableTimestamp()).thenReturn(ONE);
+            when(timelockService.getFreshTimestamp()).thenReturn(TEN);
+            when(cleaner.getUnreadableTimestamp()).thenReturn(FORTY_TWO);
+
+            assertThat(getGauge(IMMUTABLE_TIMESTAMP_NAME).getValue()).isEqualTo(ONE);
+            assertThat(getGauge(FRESH_TIMESTAMP_NAME).getValue()).isEqualTo(TEN);
+            assertThat(getGauge(UNREADABLE_TIMESTAMP_NAME).getValue()).isEqualTo(FORTY_TWO);
+
+            verify(timelockService).getImmutableTimestamp();
+            verify(timelockService).getFreshTimestamp();
+            verify(cleaner).getUnreadableTimestamp();
+            verifyNoMoreInteractions(timelockService, cleaner);
```

Comment:
This could be split into three tests, or add that `verify(timelockService).times(1).getImmutableTimestamp()` - i.e. we don't get the immutable timestamp when getting the fresh timestamp metric.

Reply:
No harm splitting this into 3 tests, which I think would read much better.

I guess you mean after the relevant statements? There's technically a funky edge case, though, where if `Immutable` called both immutable and fresh, and `Fresh` called nothing, where the test would still pass. We could of course stop that with a `verifyNoMoreInteractions()` but by then we've basically written three tests.

Reply:
I meant to add the three verifies after the three asserts, but good point about the funky case.  The three tests look much better!